src/lib/comms/sol-http-server-impl-microhttpd.c

```diff
@@ -0,0 +1,587 @@
+        FD_SET(fd, &rs);
+    if (flags & SOL_FD_FLAGS_OUT)
+        FD_SET(fd, &ws);
+    if (flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_NVAL | SOL_FD_FLAGS_HUP))
+        FD_SET(fd, &es);
+
+    /* This is far from optimized, the mainloop is given one file descriptor
+       per time while it's possible to pass all the file descriptors with events
+       to MHD at once. */
+    if (MHD_run_from_select(server->daemon, &rs, &ws, &es) == MHD_NO)
+        SOL_WRN("Something wrong happened inside microhttpd");
+
+    if (flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_NVAL | SOL_FD_FLAGS_HUP)) {
+        SOL_VECTOR_FOREACH_IDX (&server->fds, connection, i) {
+            if (connection->fd == fd) {
+                sol_vector_del(&server->fds, i);
+                break;
+            }
+        }
+
+        return false;
+    }
+
+    return true;
+}
+
+static void
+notify_connection_cb(void *data, struct MHD_Connection *connection, void **socket_data,
+    enum MHD_ConnectionNotificationCode code)
+{
+    static const enum sol_fd_flags fd_flags =
+        SOL_FD_FLAGS_IN | SOL_FD_FLAGS_OUT | SOL_FD_FLAGS_ERR |
+        SOL_FD_FLAGS_HUP | SOL_FD_FLAGS_NVAL;
+    const union MHD_ConnectionInfo *info;
+    struct sol_http_server *server = data;
+    struct http_connection *conn;
+
+    if (code != MHD_CONNECTION_NOTIFY_STARTED)
+        return;
+
+    info = MHD_get_connection_info(connection, MHD_CONNECTION_INFO_CONNECTION_FD);
+    SOL_NULL_CHECK(info);
+
+    conn = sol_vector_append(&server->fds);
+    SOL_NULL_CHECK(conn);
+
+    conn->fd = (int)info->connect_fd;
+    conn->watch = sol_fd_add(conn->fd, fd_flags, connection_watch_cb, server);
+
+    if (!conn->watch) {
+        SOL_WRN("Could not create the fd watch");
+        sol_vector_del(&server->fds, server->fds.len - 1);
+    }
+}
+
+static void
+free_request(struct sol_http_request *request)
+{
+    struct sol_http_param_value *value;
+    uint16_t idx;
+
+    if (request->pp)
+        MHD_destroy_post_processor(request->pp);
+
+    SOL_HTTP_PARAM_FOREACH_IDX (&request->params, value, idx) {
+        switch (value->type) {
+        case SOL_HTTP_PARAM_POST_FIELD:
+        case SOL_HTTP_PARAM_HEADER:
+        case SOL_HTTP_PARAM_COOKIE:
+            free((char *)value->value.key_value.value);
+            free((char *)value->value.key_value.key);
+            break;
+        default:
+            break;
+        }
+    }
+
+    sol_http_param_free(&request->params);
+}
+
+static void
+notify_connection_finished_cb(void *data, struct MHD_Connection *connection,
+    void **con_data, enum MHD_RequestTerminationCode code)
+{
+    struct sol_http_server *server = data;
+    struct sol_http_request *itr, *request = *con_data;
+    uint16_t idx;
+
+    if (!request)
+        return;
+
+    free_request(request);
+    SOL_VECTOR_FOREACH_IDX (&server->requests, itr, idx) {
+        if (itr == request) {
+            sol_vector_del(&server->requests, idx);
+            break;
+        }
+    }
+}
```

Comment:
Should a warning be emitted if the request was not found in the request vector, for debugging purposes?